crates/meilisearch/tests/upgrade/v1_12/v1_12_0.rs

```diff
@@ -0,0 +1,268 @@
+        master_key: Some("kefir".to_string()),
+        ..default_settings(temp.path())
+    };
+    copy_dir_all(original_db_path, &options.db_path).unwrap();
+    let mut server = Server::new_with_options(options).await.unwrap();
+    server.use_api_key("kefir");
+
+    check_the_keys(&server).await;
+    check_the_index_scheduler(&server).await;
+    check_the_index_features(&server).await;
+}
+
+/// We must ensure that the keys database is still working:
+/// 1. Check its content
+/// 2. Ensure we can still query the keys
+/// 3. Ensure we can still update the keys
+async fn check_the_keys(server: &Server) {
+    // All the api keys are still present
+    let (keys, _) = server.list_api_keys("").await;
+    snapshot!(json_string!(keys, { ".results[].updatedAt" => "[date]" }), name: "list_all_keys");
+
+    // We can still query the keys
+    let (by_uid, _) = server.get_api_key("9a77a636-e4e2-4f1a-93ac-978c368fd596").await;
+    let (by_key, _) = server
+        .get_api_key("760c6345918b5ab1d251c1a3e8f9666547628a710d91f6b1d558ba944ef15746")
+        .await;
+
+    assert_eq!(by_uid, by_key);
+    snapshot!(json_string!(by_uid, { ".updatedAt" => "[date]" }), @r#"
+    {
+      "name": "Kefir",
+      "description": "My little kefirino key",
+      "key": "760c6345918b5ab1d251c1a3e8f9666547628a710d91f6b1d558ba944ef15746",
+      "uid": "9a77a636-e4e2-4f1a-93ac-978c368fd596",
+      "actions": [
+        "stats.get",
+        "documents.*"
+      ],
+      "indexes": [
+        "kefir"
+      ],
+      "expiresAt": null,
+      "createdAt": "2025-01-16T14:43:20.863318893Z",
+      "updatedAt": "[date]"
+    }
+    "#);
+
+    // Remove a key
+    let (_value, status) = server.delete_api_key("9a77a636-e4e2-4f1a-93ac-978c368fd596").await;
+    snapshot!(status, @"204 No Content");
+
+    // Update a key
+    let (value, _) = server
+        .patch_api_key(
+            "dc699ff0-a053-4956-a46a-912e51b3316b",
+            json!({ "name": "kefir", "description": "the patou" }),
+        )
+        .await;
+    snapshot!(json_string!(value, { ".updatedAt" => "[date]" }), @r#"
+    {
+      "name": "kefir",
+      "description": "the patou",
+      "key": "4d9376547ed779a05dde416148e7e98bd47530e28c500be674c9e60b2accb814",
+      "uid": "dc699ff0-a053-4956-a46a-912e51b3316b",
+      "actions": [
+        "search"
+      ],
+      "indexes": [
+        "*"
+      ],
+      "expiresAt": null,
+      "createdAt": "2025-01-16T14:24:46.264041777Z",
+      "updatedAt": "[date]"
+    }
+    "#);
+
+    // Everything worked
+    let (keys, _) = server.list_api_keys("").await;
+    snapshot!(json_string!(keys, { ".results[].updatedAt" => "[date]" }), name: "list_all_keys_after_removing_kefir");
+}
+
+/// We must ensure the index-scheduler database is still working:
+/// 1. We can query the indexes and their metadata
+/// 2. The upgrade task has been spawned and has been processed (wait for it to finish or it'll be flaky)
```

Comment:
a way for the test not to be flaky and have unprocessed task would be to test the case of a failing upgrade.

Properties to check:

1. After the upgrade task fails, no task is ever processed
2. The VERSION file has been rewritten
3. After the upgrade task fails, it is still possible to enqueue new tasks
4. Restarting the server after the task failed restart the task and attempts it again

- Testing (1), (2), (3) can be done by doctoring a data.ms to contain e.g. an index with the wrong version that will cause the upgrade to fail.
- Testing (4) can be done by doctoring a data.ms to contain a failed upgrade task (for example, as in the previous bullet point), then doctoring again such that the task will pass (for example, by removing the wrong version in the faulty index)